src/soep_preparation/dataset_merging/task.py

```diff
@@ -33,216 +58,147 @@ def _dataset_level(dataset: pd.DataFrame) -> str:
+    name_to_variables_mapping: dict[str, list],
+    how: str = "outer",
+) -> pd.DataFrame:
+    out = None
+    for name, variables in name_to_variables_mapping.items():
+        columns = [
+            var for var in MERGE_KEYS if var in name_to_df_mapping[name].columns
+        ] + variables
+        if out is None:
+            out = name_to_df_mapping[name][columns]
         else:
-            datasets_by_level[level][name] = dataset
-    # Change order of datasets to make sure datasets with most reliable information on
-    # hh_id are merged first
-    datasets_by_level["individual-time-varying"] = _reorder_dict(
-        datasets_by_level["individual-time-varying"],
-        ["pequiv", "pgen", "pl"],
-    )
-    return datasets_by_level
-
+            shared_columns = _get_shared_columns(out, name_to_df_mapping[name])
+            right = name_to_df_mapping[name][columns]
+            out = out.merge(right, on=shared_columns, how=how)
+    return out
 
-def _add_origin_dummy(
-    datasets: dict[str, dict[str, pd.DataFrame]],
-) -> dict[str, pd.DataFrame]:
-    for dataset_group in datasets:
-        for dataset in datasets[dataset_group]:
-            datasets[dataset_group][dataset][f"from_{dataset}"] = pd.Series(
-                [1] * len(datasets[dataset_group][dataset]),
-                dtype="uint8[pyarrow]",
-            )
-    return datasets
 
-
-def combine_similar_information_from_different_columns(
-    data_merged: pd.DataFrame,
+def merge_data(
+    datasets: list[pd.DataFrame],
+    how: str = "outer",
 ) -> pd.DataFrame:
-    """Combining variables from different datasets with similar information.
-
-    Args:
-        data_merged (pd.DataFrame): Merged dataset.
+    out = None
+    for df in datasets:
+        if out is None:
+            out = df
+        elif df is not None:
+            shared_columns = _get_shared_columns(out, df)
+            out = out.merge(df, on=shared_columns, how=how)
+    return out
+
+
+def dataset_creation(
+    variables: list[str],
+    min_survey_year: int,
+    max_survey_year: int,
+    merging_behavior: str = "auto",
+) -> pd.DataFrame:
+    # TODO (@felixschmitz): add error handling
+    # Check if the provided variable names are valid type
+    # Check if the provided variable names are present in any dataset
+    # Check if the provided min/max survey years are valid
+    # Check if the provided merging behavior is supported
+
+    # TODO (@hmvgaudecker): what do you think?
+    # there are two variables with identical information, but potentially different values
+    # - 'birth_month_from_bioedu', 'birth_month_from_ppathl'; previously NA values in the first were filled with the second, favoring this since the first one does not have a time component
+    # - 'hh_soep_sample_from_design', 'hh_soep_sample_from_hpathl'; previously NA values in the first were filled with the second, same argument as above
+
+    name_to_df_mapping = {
```

Comment:
I think all of this should come out of the previous tasks in the pipeline. In particular, I think we want to have meta-information on each dataset. (Likely incomplete) list:

- Index columns (=potential merge keys)
- All column names
- Data types for each column

This could be two nested dictionaries with maps (abusing annotation syntax):
1. `dataset: list[merge keys]`
2. `dataset: dict[column: dtype]`

For this step, it will then be very easy to reverse the mapping in 2. so we just have `column: dataset`; for documentation on the website we likely want to have only `column: dtype`, etc.

With 1., we can get rid of the globals in this module.

Reply:
I gather you implemented much of this; thanks a lot! 

However, it is not reflected yet in the README / any docs, so I find it hard to put the pieces together. See my overall comment that will follow in a couple of minutes.